src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+        _, shift_phases, h = self._generate_shift_phase_and_filter(r_max, max_shift, shift_step)
+
+        d_theta = np.pi / n_theta
+
+        # Generate two index lists for [i, j] pairs of images
+        idx_i = []
+        idx_j = []
+        for i in range(n_img-1):
+            tmp_j = range(i + 1, n_img)
+            idx_i.extend([i] * len(tmp_j))
+            idx_j.extend(tmp_j)
+        idx_i = np.array(idx_i, dtype='int')
+        idx_j = np.array(idx_j, dtype='int')
+        # Select random pairs based on the size of n_equations
+        rp = np.random.choice(np.arange(len(idx_j)), size=n_equations, replace=False)
+
+        # Go through all shift equations in the size of n_equations
+        for shift_eq_idx in range(n_equations):
+            i = idx_i[rp[shift_eq_idx]]
+            j = idx_j[rp[shift_eq_idx]]
+            # get the common line indices based on the rotations from i and j images
+            c_ij, c_ji = self._get_cl_indices(rotations, i, j, n_theta)
+
+            # check whether need to flip or not
+            is_pf_j_flipped = 0
+            if c_ji < n_theta:
+                pf_j = pf[:, c_ji, j]
+            else:
+                pf_j = pf[:, c_ji - n_theta, j]
+                is_pf_j_flipped = 1
+            pf_i = pf[:, c_ij, i]
+
+            # perform bandpass filter, normalize each ray of each image,
+            # and only keep half of each ray
+            pf_i, pf_j = self._apply_filter_and_norm(pf_i, pf_j, r_max, h)
+
+            # apply the shifts to images
+            pf_i_flipped = np.conj(pf_i)
+            pf_i_stack = np.einsum('i, ij -> ij', pf_i, shift_phases)
+            pf_i_flipped_stack = np.einsum('i, ij -> ij', pf_i_flipped, shift_phases)
+
+            c1 = 2 * np.real(np.dot(np.conj(pf_i_stack.T), pf_j))
+            c2 = 2 * np.real(np.dot(np.conj(pf_i_flipped_stack.T), pf_j))
+
+            # find the indices for the maximum values
+            # and apply corresponding shifts
+            sidx1 = np.argmax(c1)
+            sidx2 = np.argmax(c2)
+
+            if c1[sidx1] > c2[sidx2]:
+                dx = -max_shift + sidx1 * shift_step
+            else:
+                dx = -max_shift + sidx2 * shift_step
+
+            idx = np.arange(4 * shift_eq_idx, 4 * shift_eq_idx + 4)
+            shift_alpha = c_ij * d_theta
+            shift_beta = c_ji * d_theta
+            shift_i[idx] = shift_eq_idx
+            shift_j[idx] = [2 * i, 2 * i + 1, 2 * j, 2 * j + 1]
+            shift_b[shift_eq_idx] = dx
+
+            # check with compare cl
+            if not is_pf_j_flipped:
+                shift_eq[idx] = [np.sin(shift_alpha), np.cos(shift_alpha),
+                                 -np.sin(shift_beta), -np.cos(shift_beta)]
+            else:
+                shift_beta -= np.pi
+                shift_eq[idx] = [-np.sin(shift_alpha), -np.cos(shift_alpha),
+                                 -np.sin(shift_beta), -np.cos(shift_beta)]
+
+        t = 4 * n_equations
```

Comment:
The last part is used to store the information of right side (`shift_b`) so that all information can be saved in the `shift_equations` object. Since we also save shift_b, this part was removed in the new commit b9a012a.